### PR TITLE
bring back the takes UI

### PR DESCRIPTION
**Pull request by whit537**
&#x2B11; #3994

&larr; #4099 ([diff](https://github.com/gratipay/gratipay.com/compare/update-some-python-APIs...preen-takes))

Refactor the new takes/membership code now that #4023 and #4024 are merged into this new line of development.

- [x] rebase once #4076 lands
- [x] don't call `get_take_for` from `set_take_for` (use `old_takes`)
- [x] deduplicate error handling between python and simplates
- [x] rebase once #4090 lands
- [x] prevent owner from joining a Team (new PR?)
- [x] factor out a few other PRs ... #4092 #4093 #4094 #4099
- [x] rebase once #4099 lands
- [x] restrict the distributing page—https://github.com/gratipay/gratipay.com/pull/4077#discussion_r73420701 —done in d39d41101066feb85dc88a38f53eafc88f1c392c
- [x] _don't_ restrict the distributing page :) —https://github.com/gratipay/gratipay.com/pull/4077#discussion_r73428674 —reverted in 1733e21a7dd15f9158d088edcb705927ea75dda9
- [x] back out the `lookup.json` l10n—done in 505c14f34ece459eabc3cb759ead21062205209e
- [x] back out the admin bar l10n— https://github.com/gratipay/gratipay.com/pull/4077#discussion_r73401849 —done in 064b88c70463a28fc3457ea67c1e7b4760099b0e
- [ ] cram the table into the layout—https://github.com/gratipay/gratipay.com/pull/4077#discussion_r73421480
- [ ] clean up sidebar—https://github.com/gratipay/gratipay.com/pull/4077#discussion_r73586902
- [x] ~~nicely~~ prevent owners from adding ineligible members—https://github.com/gratipay/gratipay.com/pull/4077#discussion_r73422575 —done in c0cf218f2027c820c5cb6ee35fc3a93c96ee0bc6 and 51d9739340f18a62ca4be4cb5a8edcaf5886839d
- [x] consider fixing test helper for Chrome 52—https://github.com/gratipay/gratipay.com/pull/4077#discussion_r73425567 —done in 1866b966673278a3f1802f0938bdfa5117bf8da8 2be3a87a7feeebbc08b76bd99bdc7e828cc7c91a aa5a093fa4f5f31885cdde83e46490e7f69ad0ec and 68ac0989fae54f8f828dfbffdcc65ebb45407c60

### CODE REVIEW

**Comment by whit537** [comment]
Page is barely loading! 😱 

screen shot 2016-07-07 at 12 20 12 pm

**Comment by whit537** [comment]
Yay! 😍 

screen shot 2016-07-07 at 2 03 41 pm

**Comment by whit537** [comment]
Alright, I migrated client-side tests to Splinter in #4079, and have inserted that into the PR chain. Ready to write some tests for this UI! :-)

**Comment by whit537** [comment]
#4076 is in, and I've rebased on master. Onward!

**Comment by whit537** [comment]
BLAM!!!!!!!!!!!!!!!!!!!

Adding members, tested ... 👍 

screen shot 2016-07-14 at 12 11 01 pm
screen shot 2016-07-14 at 12 10 52 pm

**Comment by whit537** [comment]
As part of getting this to work, I made a couple other changes that should be separate PRs.

**Comment by whit537** [comment]
2af9e57202518fb7a085df3c9d10517ce0a34566 2378e26ec8dd673881ce466d78b178f5d7dbd5cc f3dc41509f0731a25561ea48f7a7241029fa67a6 could/should be cherry-picked into separate PRs.

**Comment by whit537** [comment]
Hmm ... a test failure to investigate. Perhaps related to PhantomJS vs. Chrome?

**Comment by whit537** [comment]
Okay! Travis is happy. 👍

Now to bring removing users, and then setting takes ...

**Comment by whit537** [comment]
ba81708702871c0632c553753ba7a5d699da30f8 and 6d190c527614640fb9e75c4245785409e12c47a2 can go in a separate PR as well.

**Comment by whit537** [comment]
... aaaaaaaand 1ea129496f8d959bad2dc8051582446af3e2db6f. :)

**Comment by nashe** [comment]
You won't need the lookup.json anymore? Otherwise, I can make the small adjustments we talked about (~ don't return the non-searchable users except in the case of an exact match).

**Comment by whit537** [comment]
> You won't need the lookup.json anymore? Otherwise, I can make the small adjustments we talked about (~ don't return the non-searchable users except in the case of an exact match).

We should address that before going live with the rest of the work on this PR, yes. Wanna do that in a separate PR, and we can be sure to deploy that before landing this PR?

**Comment by nashe** [comment]
I'm getting a bunch of errors while trying to run only one test. I'll finish the tests and make the PR after lunch :-)

**Comment by whit537** [comment]
> I'm getting a bunch of errors while trying to run only one test. I'll finish the tests and make the PR after lunch :-)

If you start a PR with the broken code, we can maybe help debug. :)

**Comment by whit537** [comment]
(You could be hitting problems introduced by #4079.)

**Comment by nashe** [comment]
I think I'll add a Makefile rule to be able to only run a batch of tests (pytest's `-k` option).
The PR is alive in #4090!

**Comment by whit537** [comment]
> I think I'll add a Makefile rule to be able to only run a batch of tests (pytest's `-k` option).

Go for it. I have this in my `.profile`:

```
alias envtest="honcho run -e defaults.env,local.env,tests/test.env,tests/local.env py.test"
```

So I run tests like this:

```
envtest tests/ttw/test_foo.py -k test_foo_
```

**Comment by whit537** [comment]
I'm seeing a [weird error](https://travis-ci.org/gratipay/gratipay.com/builds/146223829) from Travis for 283b5aa8b03b01aa5b00aadc3d0d9257735699e2:

> ERROR: An error occured while trying to fetch your .travis.yml file.
>
Is GitHub down? Please contact support@travis-ci.com if the error persists.

Looks like maybe https://github.com/travis-ci/travis-ci/issues/2025?

**Comment by whit537** [comment]
Blech. Same for aa9b49030670bcff65225b6ad881a343ddf1bda4: https://travis-ci.org/gratipay/gratipay.com/builds/146233269. :-/

Buzzkill!

**Comment by whit537** [comment]
To: support@travis-ci.com
Subject: error while trying to fetch .travis.yml

> Greetings! We've started seeing the following on one of our PRs:
>>
ERROR: An error occured while trying to fetch your .travis.yml file. 
>>
Is GitHub down? Please contact support@travis-ci.com if the error persists.
>
The PR doesn't touch .travis.yml:
>
```
$ git diff --name-only master | grep travis
$
```
>
And GitHub does not appear to be down.
>
How should we proceed?

**Comment by whit537** [comment]
I've rebased on `master`, but the problem with Travis [persists](https://travis-ci.org/gratipay/gratipay.com/builds/146238014).

**Comment by whit537** [comment]
I've pushed up the remainder of the commits we have for today. The test suite is passing locally with Chrome. There's one test that fails with PhantomJS, because it doesn't support `alert`s.

**Comment by whit537** [comment]
From: Travis

> Thank you for writing in and my apologies for the hassle. 
>
I've been digging through our logs and seems that while trying to fetch the _.travis.yml_ file for your repository, there was an issue due to bad credentials.
>
Upon some more investigation, seems that this was due to a permissions issue with the username **nashe**, is that possible?
>
That said, I've triggered a sync for Gratipay which should have updated the permissions, could you push a new commit to the pull request and let me know if that triggers a build successfully?
>
Thank you again and please, keep us posted - keen to figure this out for you!

**Comment by whit537** [comment]
I did a bunch of rebasing this morning on the bus, moving commits up to the front of the branch so we can make new PRs for them. I've pushed that up, so let's see if Travis works now!

**Comment by whit537** [comment]
It's [building](https://travis-ci.org/gratipay/gratipay.com/builds/146370331), so it looks like we're back on track. 👍

**Comment by whit537** [comment]
To: Travis

> Thanks, []. nashe is a new contributor, so I suppose there could be a permissions issue, though he doesn't have any commits on this PR, so I'm not sure what the specific issue could be.
>
I've pushed to the branch and we're back on track! In the future I'll keep the resync option in mind as something to try with these kind of issues.
>
Thanks for your help! Have a great day! :-)

File: js/gratipay/team.js [hunk 1]
```diff
@@ -192,6 +191,7 @@ Gratipay.team = (function() {
                     if(d.success) {
                         Gratipay.notification(d.success, 'success');
                     }
+                    debugger;
```
Comment (nashe) [inline js/gratipay/team.js]: Forgot to remove it?

Comment (whit537) [inline js/gratipay/team.js]: Heh, looks like it! Wonder why this didn't break tests?

Comment (nashe) [inline js/gratipay/team.js]: Maybe they want you to declare an remote debugger or they'll simply ignore this instruction if none is provided? Can't find any documentation about this. I'll try by myself once @home :-)

Comment (whit537) [inline js/gratipay/team.js]: Could be. I know that Chrome ignores `debugger` statements when the developer tools are not open.

Comment (whit537) [inline js/gratipay/team.js]: Removed in 9287109e04b4545cf586ecccc5d2f59938ccc51e. I'll probably squash it with the commit that introduced it.

**Comment by whit537** [comment]
From: Travis

> Thank you for the follow-up, I'm happy to hear that it's working as expected now. 
>
To avoid this situation from happening again in the future, I'd recommend triggering a sync at both https://travis-ci.org/profile/gratipay and at the profile of the new contributor.
>
You know we're here if you run into something else :)

**Comment by whit537** [comment]
Okay! Let's see about landing #4090, and then queueing up some additional PRs from this branch.

**Comment by whit537** [comment]
Rebased to pick up #4090. Now to carve off some new PRs ...

**Comment by whit537** [comment]
Rebased post-#4099! Previous head was 0006e626bd05bcd79396b20b4ec4d3a2a09425eb. Eeeeeee!

**Comment by whit537** [comment]
@kaguillera and I were going back and forth last time we were together on the question of owners adding themselves to a team. Since it's more work to prevent than to allow, and there's no obvious danger to it (right?), I say we led it slide: owners can add themselves to the team if they want. That could enable more-or-less what we had under 1.0, where the owner was an individual taking from the team just like everyone else.

**Comment by whit537** [comment]
Yeesh! Is this ready for final review? 😳

Once we merge this, we'll have UI to add participants as team members! Then we'll just need #4025 (or a redo) in order to close #3994! 🙀

**Comment by nashe** [comment]
> where the owner was an individual taking from the team just like everyone else.

_(From the point of view of somebody that never used the 1.0)_, it seems to be the best idea.

File: www/%team/distributing/%to.json.spt [hunk 2]
```diff
@@ -9,87 +9,77 @@ from gratipay.models.participant import Participant
...
     raise Response(404, _("User not found."))
-is_member = member.member_of(team)
 
-if request.method == 'POST':
+participant_is_member = participant.member_of(team)
+recorder_is_owner = team.owner == recorder.username
 
-    if user.participant not in (member, team):
-        raise Response(403)
+if request.method == 'POST':
```
Comment (nashe) [inline www/%team/distributing/%to.json.spt]: Don't we already have https://github.com/gratipay/gratipay.com/pull/4077/commits/a7c5fae31432896ca3597310facf79e8e69e8aff#diff-0f0c9aee41837b89ae6ba33b2e09965bR12 ?

**Comment by whit537** [comment]
This plays into (the long and rambling) https://github.com/gratipay/inside.gratipay.com/issues/432. For projects that don't have a legal entity, it can make sense for the owner to be on the distribution list alongside everyone else. Where the owner is a legal entity, that could make less sense.

File: www/%team/distributing/%to.json.spt (hunk 2 above)
Comment (whit537) [inline www/%team/distributing/%to.json.spt]: Good point.

Comment (whit537) [inline www/%team/distributing/%to.json.spt]: Addressed in d7a4be80a33c1d82f2922c14f81a5738f74fee4b.

**Comment by nashe** [comment]
!m @whit537 :-)

**Comment by mattbk** [comment]
Nice work.

Doesn't the owner end up with the excess from takes anyway, in order to pay the bills? Could the owner take a certain amount (as an individual) and then the excess is a separate line item on payday, which makes it easy for the owner to distinguish the accounts?

**Comment by whit537** [comment]
@mattbk Thanks! :)

Yeah, that's the sort of thing I was thinking. Examples of the two patterns:

1. The ~owner of [**DevDocs**](https://gratipay.com/devdocs/) is [~Thibaut](https://gratipay.com/~Thibaut/). ~Thibaut could _also_ add himself as a _member_ of devdocs, to differentiate the amount he (and his collaborators) view as his salary equivalent, and the amount that is for the coffers of the team itself. Both are going to him, but we can help him with the accounting.
1. The ~owner of [**Gratipay**](https://gratipay.com/Gratipay/) is [~Gratipay](https://gratipay.com/~Gratipay/) (cf. #3852). We'll have to agree on an amount to set as our weekly _available_ for distribution—[maybe](https://github.com/gratipay/inside.gratipay.com/issues/747) $50 to start with? ~Gratipay won't be a member of **Gratipay**. ~Gratipay will receive a) any income above the (e.g.) $50 set as "available", plus b) anything left over from the $50 that we members don't take.

**Comment by whit537** [comment]
@mattbk Does that fit with how you're thinking about it?

**Comment by nashe** [comment]
The comments I have so far (it's starting to be late here so I'll dig into this tomorrow): 
- The menu on the left linking to "Distributing" is only displayed when the user is admin (not for the team owner) and there is no other reference to this page—I don't really know what the team interface is supposed to look like but I guess that [this condition](https://github.com/gratipay/gratipay.com/blob/a52c0f958ec362294a9214f768d52ed92b8afd7a/templates/team-base.html#L56) is wrong: we should use `user.participant.username` instead, no? 
- The "Member" word is not localized. Nor the administration bar (belongs to another PR).
- The last column of the table breaks the design :-( But we need all the data that is inside this table.
- It's easy to trigger `assert p.has_verified_identity, p.id`, should we change `lookup.json` to display only the verified or return a nicer error message?

**Comment by nashe** [comment]
(updated my last message).

**Comment by whit537** [comment]
@nashe Make 'em [todos](https://github.com/blog/1375-task-lists-in-gfm-issues-pulls-comments) so we can check them off! 💃

**Comment by mattbk** [comment]
> Does that fit with how you're thinking about it?

Yes, exactly. The way it is now, teams have a choice of how they want to do the accounting (your two scenarios).

**Comment by whit537** [comment]
Cool, let's roll with that, then, if no-one else objects ...

**Comment by whit537** [comment]
Actually, I just noticed while working on #4025 that owners can join a team, but are currently prevented from upping their take beyond a penny.

```
================================================== FAILURES ==================================================
______________________________ TestTakes.test_pt_is_happy_to_deal_the_owner_in _______________________________
Traceback (most recent call last):
  File "/Users/whit537/personal/gratipay/gratipay.com/tests/py/test_billing_payday.py", line 645, in test_pt_is_happy_to_deal_the_owner_in
    self.enterprise.set_take_for(self.picard, 50, self.picard)
  File "/Users/whit537/personal/gratipay/gratipay.com/gratipay/models/team/mixins/takes.py", line 84, in set_take_for
    raise NotAllowed('owner can only add and remove members, not otherwise set takes')
NotAllowed: owner can only add and remove members, not otherwise set takes
==================================== 31 tests deselected by '-ktest_pt_' =====================================
============================= 1 failed, 3 passed, 31 deselected in 5.36 seconds ==============================
[gratipay] $
```

**Comment by whit537** [comment]
So we'll have to relax that constraint.

**Comment by nashe** [comment]
> Actually, I just noticed while working on #4025 that owners can join a team, but are currently prevented from upping their take beyond a penny.

Yep, that's on the todolist of my last message! Should I put it elsewhere? I'll take a look at the localization part first. Can you confirm what I said for the first point?

**Comment by whit537** [comment]
I'm not sure if we're talking about the same issue, but having it on the todo list in your last message is fine. Re: the first point, you may be right. Maybe write a test and see if it fails? :)

I am heading out for the weekend, I'll probably check in again on Sunday evening. TTFN! ⏩

**Comment by nashe** [comment]
Hum should I directly push it here or make a PR?

**Comment by whit537** [comment]
Where do we stand here? I see a few things crossed off on the [list](https://github.com/gratipay/gratipay.com/pull/4077#issuecomment-236036025) but no further commits pushed. Do you have some changes locally waiting on the answer to https://github.com/gratipay/gratipay.com/pull/4077#issuecomment-236449311?

I'm ready to work on this this afternoon. I would say push whatever commits you have and I will pick up in a bit after lunch! :-)

**Comment by nashe** [comment]
Regarding "Should the distributing page be restricted?", I'd say yes, restricted to the team members. Anyway, the anonymous/non-members users don't see the link, so I guess that it's what you want. Just let me know if it's the case or not!

File: gratipay/models/team/mixins/takes.py [hunk 3]
```diff
@@ -3,6 +3,7 @@
 from collections import OrderedDict
 from decimal import Decimal as D
 
+from gettext import gettext as _
```
Comment (whit537) [inline gratipay/models/team/mixins/takes.py]: Are you sure we import from `gettext`? I think we have some wrappers that we use instead of raw `gettext`, need to check ...

File: gratipay/models/team/mixins/takes.py [hunk 4]
```diff
@@ -71,19 +72,23 @@ def set_take_for(self, participant, take, recorder, cursor=None):
 
         """
         def vet(p):
-            assert not p.is_suspicious, p.id
-            assert p.is_claimed, p.id
-            assert p.email_address, p.id
-            assert p.has_verified_identity, p.id
+            if p.is_suspicious:
+                raise NotAllowed(_("user must not be flagged as suspicious"))
```
Comment (whit537) [inline gratipay/models/team/mixins/takes.py]: Do we actually surface these messages in the UI?

File: templates/team-base.html [hunk 5]
```diff
@@ -39,21 +39,21 @@
 <form id="status-knob" action="set-status.json">
     <select name="status">
         <option {{ 'selected' if team.is_approved == None else '' }}
-            value="unreviewed">Under Review</option>
+            value="unreviewed">{{ _("Under Review") }}</option>
```
Comment (whit537) [inline templates/team-base.html]: To date, we haven't worried about translating our admin interface. Seems like scope creep to do so on this PR, no?

Comment (whit537) [inline templates/team-base.html]: Heh. :-)

> Nor the administration bar (belongs to another PR).

https://github.com/gratipay/gratipay.com/pull/4077#issuecomment-236036025

File: gratipay/models/team/mixins/takes.py (hunk 3 above)
Comment (nashe) [inline gratipay/models/team/mixins/takes.py]: I did not found it, I'll try to fix this.

File: gratipay/models/team/mixins/takes.py (hunk 4 above)
Comment (nashe) [inline gratipay/models/team/mixins/takes.py]: Right, we need to display this.

File: templates/team-base.html (hunk 5 above)
Comment (nashe) [inline templates/team-base.html]: 😇

File: gratipay/models/team/mixins/takes.py (hunk 4 above)
Comment (whit537) [inline gratipay/models/team/mixins/takes.py]: Hmm ... I was thinking that the UI should prevent users from ever hitting these messages in the first place, as you [suggest](https://github.com/gratipay/gratipay.com/pull/4077#issuecomment-236036025) with precluding non-verified, etc. users from `lookup.json`.

Comment (whit537) [inline gratipay/models/team/mixins/takes.py]: That way, you'd only hit this if you were hitting the API directly. At that point I think English error messages are okay.

**Comment by whit537** [comment]
@nashe In future let's try asking question in line comments on the diffs, and then linking to those from a Todo in the ticket description (cf. the much simpler #4097).

The problem is that I want to respond to points you made in https://github.com/gratipay/gratipay.com/pull/4077#issuecomment-236036025, but it's hard to do so and keep the various threads of conversation somewhat organized. :)

File: www/%team/distributing/index.html.spt [hunk 6]
```diff
@@ -36,25 +36,25 @@ title = _("Team Members")
...
+            <ul class="lookup-results"></ul>
           </form>
         </div>
       </td>
     </tr>
     {% endif %}
     </thead>
     <tfoot><tr><td colspan=6><div class="loading-indicator"/></td></tr></foot>
-    <tbody id="team-members"></tbody>
+    <tbody class="team-members"></tbody>
```
Comment (whit537) [inline www/%team/distributing/index.html.spt]: @nashe at https://github.com/gratipay/gratipay.com/pull/4077#issuecomment-236036025:

> Should the distributing page be restricted?

File: js/gratipay/team.js [hunk 7]
```diff
@@ -61,43 +56,50 @@ Gratipay.team = (function() {
...
+        } else {
+            rows.push(Gratipay.jsonml(
+                [ 'tr'
+                , {'class': 'totals'}
+                , ['td']
+                , ['td']
+                , ['td']
+                , ['td', {'class': 'figure take'}, num(available - member.balance)]
+                , ['td', {'class': 'figure balance'}, num(member.balance)]
+                , ['td', {'class': 'figure percentage'}, perc(member.balance / available)]
```
Comment (whit537) [inline js/gratipay/team.js]: @nashe at https://github.com/gratipay/gratipay.com/pull/4077#issuecomment-236036025:

> The last column of the table breaks the design :-( But we need all the data that is inside this table.

File: templates/team-base.html (hunk 5 above)
Comment (nashe) [inline templates/team-base.html]: Reverted few seconds ago.

File: js/gratipay/team.js [hunk 8]
```diff
@@ -61,43 +56,50 @@ Gratipay.team = (function() {
...
     // ===
 
     function lookup() {
-        var query = $('#query').val();
+        var query = $t('.query').val();
         if (query === '')
-            $('#lookup-results').empty();
+            $t('.lookup-results').empty();
         else
             jQuery.get("/lookup.json", {query: query}).success(drawLookupResults);
```
Comment (whit537) [inline js/gratipay/team.js]: As @nashe points out in https://github.com/gratipay/gratipay.com/pull/4077/files#r73421480, this endpoint returns users that don't pass the validation for new team members. In e80433913ec6030987c9a9007d16050432fb83d6 we have a start on wiring up proper error messaging, but I think it'd be better to go the alternate route that @nashe suggests, and not include these non-allowed users in the first place. The best would probably be to include them in `lookup.json` results, but with a flag so that we can show in the UI that "this person exists—you got their username right—but they're not eligible for such-and-such reason."

File: gratipay/models/team/mixins/takes.py (hunk 4 above)
Comment (nashe) [inline gratipay/models/team/mixins/takes.py]: Should I just revert the l10n?

**Comment by whit537** [comment]
064b88c70463a28fc3457ea67c1e7b4760099b0e 👍 

I've taken your comments in https://github.com/gratipay/gratipay.com/pull/4077#issuecomment-236036025 and reposted them on the diff. I've updated the main PR description with a "Todo" that links out to these reposts for additional discussion and decision.

File: www/%team/distributing/index.html.spt (hunk 6 above)
Comment (whit537) [inline www/%team/distributing/index.html.spt]: I don't think it should be, and certainly not by default. Why _would_ we restrict it?

File: gratipay/models/team/mixins/takes.py (hunk 3 above)
Comment (whit537) [inline gratipay/models/team/mixins/takes.py]: This question goes away if we go with preventing adding in the first place as a solution to the UX problem with adding invalid users (https://github.com/gratipay/gratipay.com/pull/4077#discussion_r73422575).

File: gratipay/models/team/mixins/takes.py (hunk 4 above)
Comment (whit537) [inline gratipay/models/team/mixins/takes.py]: Yeah, I think so. I'll work on `lookup.json` ...

File: gratipay/testing/browser.py [hunk 9]
```diff
@@ -80,6 +81,32 @@ def has_element(self, selector, timeout=None):
         """
         return self.is_element_present_by_css(selector, timeout)
 
+    def confirming(self, answer):
+        """Return a context manager for working with confirm alerts cross-driver.
+        """
```
Comment (whit537) [inline gratipay/testing/browser.py]: Sadly, I think that Chrome 52 may actually break this. I'm almost positive it was working under Chrome 51. :-/

File: gratipay/models/team/mixins/takes.py (hunk 4 above)
Comment (nashe) [inline gratipay/models/team/mixins/takes.py]: Reverting it…

File: www/%team/distributing/index.html.spt [hunk 10]
```diff
@@ -1,10 +1,19 @@
...
 team = get_team(state)
 if team.available == 0:
     website.redirect('..', base_url='')
+
+if user.participant == None:
+    raise Response(401)
+
+if not (user.ADMIN or user.participant.member_of(team) or user.participant.username == team.owner):
+    raise Response(403)
+
```
Comment (whit537) [inline www/%team/distributing/index.html.spt]: Sorry (and I tried to get ahead of this at https://github.com/gratipay/gratipay.com/pull/4077#discussion_r73424777), but I don't think we should restrict the distributing page. Why _would_ we?

File: js/gratipay/team.js (hunk 8 above)
Comment (whit537) [inline js/gratipay/team.js]: Done in c0cf218f2027c820c5cb6ee35fc3a93c96ee0bc6 and 51d9739340f18a62ca4be4cb5a8edcaf5886839d for the simple case (excluding invalid users from lookup.json). This PR is getting big and we're not making this feature generally available yet, only for Team Gratipay. I think we should punt on nicer handling here.

File: www/%team/distributing/index.html.spt (hunk 6 above)
Comment (whit537) [inline www/%team/distributing/index.html.spt]: Picking up at https://github.com/gratipay/gratipay.com/pull/4077#discussion_r73428674 ...

**Comment by whit537** [comment]
I'm outta time for today. I see three items still open on the todo. I should be around to work on this again tomorrow ... hopefully we can land it! :-)

File: www/%team/distributing/index.html.spt (hunk 10 above)
Comment (nashe) [inline www/%team/distributing/index.html.spt]: Oh I missed your comment… I was thinking that since there were no link to this page for the ordinary user, it was not supposed to be reachable.

Comment (nashe) [inline www/%team/distributing/index.html.spt]: Reverting it… so much reverts today, sorry!

Comment (whit537) [inline www/%team/distributing/index.html.spt]: Thanks for being a good sport, and biased towards action! :-)

Comment (whit537) [inline www/%team/distributing/index.html.spt]: > I was thinking that since there were no link to this page for the ordinary user, it was not supposed to be reachable.

It's supposed to be reachable, this is just another case where we're only making this available for Team Gratipay right now. Actually, we should still make it generally available (the Team Gratipay distributing page). Maybe we can handle this and Receiving in a new PR (cf. https://github.com/gratipay/gratipay.com/pull/3873#issuecomment-164165342)?

**Comment by whit537** [comment]
> I see three items still open on the todo.

Down to two! :-)

**Comment by whit537** [comment]
@nashe I addressed the issue with testing on Chrome 52 by implementing our own confirm dialog rather than using the native browser one. That piece is ready for review in 1866b966673278a3f1802f0938bdfa5117bf8da8 2be3a87a7feeebbc08b76bd99bdc7e828cc7c91a aa5a093fa4f5f31885cdde83e46490e7f69ad0ec and 68ac0989fae54f8f828dfbffdcc65ebb45407c60! :-)

**Comment by kaguillera** [comment]
lgtm

**Comment by nashe** [comment]
Sounds good!

**Comment by whit537** [comment]
Alright, looking at the last todo item:

> cram the table into the layout

I'm not sure why the sidebar and team name are not showing up on the page:

screen shot 2016-08-04 at 2 15 05 pm

File: js/gratipay/team.js (hunk 7 above)
Comment (whit537) [inline js/gratipay/team.js]: Hmm ... without the sidebar there's plenty of room for the table, but I think we're supposed to have the sidebar, even though we don't right now (the team name is missing as well):

screen shot 2016-08-04 at 2 15 05 pm

Comment (whit537) [inline js/gratipay/team.js]: Team name back in caccc7b3d85273fbafd4e05f761ad53f1e7a196f.

Comment (whit537) [inline js/gratipay/team.js]: The sidebar is a bit of a mess. I've refactored the tests in b6e6608f034cf90a8243b56ccd2d318b31d169d3 ahead of rationalizing the behavior here. Here's our target:

```
Profile
Receiving
Distributing

x = impossible combination

            anon    rando   member  owner   admin
unapproved                  x       
approved    P       P       PR      PR      PR
available   PD      PD      PRD     PRD     PRD
```

**Comment by whit537** [comment]
This PR is getting unwieldy. Our [rule of thumb](http://inside.gratipay.com/howto/develop-software) is that:

> The net change for a pull request (PR) shouldn't be more than about 400 lines (additions + deletions).

We're up to 436 + 233 = 669 here. I propose that we merge this PR and reticket the two remaining todo items.

File: www/%team/distributing/%to.json.spt [hunk 11]
```diff
@@ -0,0 +1,84 @@
...
+
+team = get_team(state)
+if team.available == 0:
+    website.redirect('..', base_url='')
+
+if user.ANON:
+    raise Response(401)
+recorder = user.participant
+
+participant = Participant.from_id(request.path['to'])
```
Comment (aandis) [inline www/%team/distributing/%to.json.spt]: Shouldn't this be `from_username`?

Comment (aandis) [inline www/%team/distributing/%to.json.spt]: Ah! Just saw https://github.com/gratipay/gratipay.com/pull/4077/files#diff-a4f6262e5582179057579d255f8d1d8bR129. Branch wasn't updated. Looks great now!

**Comment by aandis** [comment]
The `giving` sidebar on the `distributing` page doesn't work. Haven't been part of all the conversation but I am guessing this is a bug. Fixing..

**Comment by aandis** [comment]
Seeing some weird notification behavior on member remove. If I click on remove member and then cancel `n` times and then finally remove, notifications show up `n+1` times.

**Comment by aandis** [comment]
Looks like the endpoint is being called multiple times. https://github.com/gratipay/gratipay.com/pull/4077/files#diff-a4f6262e5582179057579d255f8d1d8bR185 could be the possible culprit.

**Comment by aandis** [comment]
Oh I get it. https://github.com/gratipay/gratipay.com/pull/4077/files#diff-591777796ffbdb67f591b255039783ebR5 registers multiple events on the same element.

**Comment by aandis** [comment]
Fixed in ef66e066e9d0d13ce89b9ece70e5fd6d28eed8f5

**Comment by whit537** [comment]
!m @aandis @nashe 

Rebased on master. Previous head was ef66e066e9d0d13ce89b9ece70e5fd6d28eed8f5.

**Comment by whit537** [comment]
Ready?!!?!?!? 😳

**Comment by whit537** [comment]
I'm planning to merge and deploy this once Travis is green.

**Comment by whit537** [comment]
💃
